service.onmute/service.py

```diff
@@ -0,0 +1,215 @@
+    if isinstance(string, basestring):
+        if isinstance(string, unicode):
+           string = string.encode('utf-8', 'ignore' )
+        else:
+           string = ascii(string)
+    return string
+     
+def loadJson(string):
+    if len(string) == 0:
+        return {}
+    try:
+        return json.loads(uni(string))
+    except:
+        return json.loads(ascii(string))
+        
+def dumpJson(mydict, sortkey=True):
+    return json.dumps(mydict, sort_keys=sortkey)
+
+def sendJSON(command):
+    data = ''
+    try:
+        data = xbmc.executeJSONRPC(uni(command))
+    except UnicodeEncodeError:
+        data = xbmc.executeJSONRPC(ascii(command))
+    return uni(data)
+    
+class Player(xbmc.Player):
+    def __init__(self):
+        xbmc.Player.__init__(self, xbmc.Player())
+        
+        
+    def onPlayBackStarted(self):
+        log('onPlayBackStarted')
+        self.subFound = False
+        
+        
+    def onPlayBackEnded(self):
+        log('onPlayBackEnded')
+        self.subFound = False
+        
+        
+    def onPlayBackStopped(self):
+        log('onPlayBackStopped')
+        self.subFound = False
+        
+        
+class Monitor(xbmc.Monitor):
+    def __init__(self):
+        xbmc.Monitor.__init__(self, xbmc.Monitor())
+        self.searchSub = REAL_SETTINGS.getSetting('searchSub') == "true"
+        self.enableCC =  REAL_SETTINGS.getSetting('enableCC') == "true"
+        
+        
+    def onSettingsChanged(self):
+        log("onSettingsChanged")
+        self.searchSub = REAL_SETTINGS.getSetting('searchSub') == "true"
+        self.enableCC =  REAL_SETTINGS.getSetting('enableCC') == "true"
+        
+        
+class Service():
+    def __init__(self):
+        self.userCC = self.getCC()
+        self.Player = Player()
+        self.Monitor = Monitor()
+        self.autoSub = False
+        self.Player.subFound = False
+        self.searchSub = self.Monitor.searchSub
+        self.start()
+
+        
+    def searchSubtitle(self):
+        log("searchSubtitle")
+        self.Player.subFound = True
+        xbmc.executebuiltin("ActivateWindow(SubtitleSearch)")
+        
+        
+    def isSubtitle(self):
+        state = xbmc.getCondVisibility('VideoPlayer.SubtitlesEnabled') == 1
+        log("isSubtitle = " + str(state))
+        return state
+         
+         
+    def setSubtitle(self, state):
+        log("setSubtitle = " + str(state))
+        self.Player.showSubtitles(state)
+         
+         
+    def hasSubtitle(self):
+        state = xbmc.getCondVisibility('VideoPlayer.HasSubtitles') == 1
+        log("hasSubtitle = " + str(state))
+        return state
+
+        
+    def isMute(self):
+        state = xbmc.getCondVisibility('Player.Muted') == 1
+        log("isMute = " + str(state))
+        return state
+
+        
+    def setMute(self, state):
+        log("setMute = " + str(state))
+        bailout = 0
+        while self.isMute() != bool(state) and bailout < 13:
```

Comment:
On paper I'd agree, but In use I've found that this condition is missable. Since the goal is to set a mute condition, it's not okay to assume it goes through. I felt a loop with a fail safe was the best option, I'm open to suggestions.